metricq_tools/summary.py

```diff
@@ -0,0 +1,294 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021, ZIH, Technische Universitaet Dresden, Federal Republic of Germany
+#
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without modification,
+# are permitted provided that the following conditions are met:
+#
+#     * Redistributions of source code must retain the above copyright notice,
+#       this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright notice,
+#       this list of conditions and the following disclaimer in the documentation
+#       and/or other materials provided with the distribution.
+#     * Neither the name of metricq nor the names of its contributors
+#       may be used to endorse or promote products derived from this software
+#       without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import asyncio
+
+import click
+import click_completion
+import click_log
+import metricq
+import numpy as np
+import termplotlib as tpl
+from metricq.drain import Drain
+from metricq.subscriber import Subscriber
+from tabulate import tabulate
+
+from metricq_tools.utils import metricq_server_option, metricq_token_option
+
+from .logging import get_root_logger
+from .version import version as client_version
+
+logger = get_root_logger()
+
+click_completion.init()
+
+
+class Summary:
+    def __init__(
+        self,
+        metrics: list[str],
+        intervals_histogram: bool,
+        values_histogram: bool,
+        print_stats: bool,
+        print_data: bool,
+    ):
+        self._metrics = metrics
+
+        self.print_intervals = intervals_histogram
+        self.print_values = values_histogram
+        self.print_stats = print_stats
+        self.print_data = print_data
+
+        self.timestamps = dict[str, list]()
```

Comment:
Instead of using one `dict` per data, wouldn't it be a good a idea to group stuff by metric?  Something like:

```python

from dataclasses import dataclass

...

@dataclass
class MetricData:
    timestamps: list[Timestamp] = []
    last_timestamp: Optional[Timestamp] = None
    intervals: list[?] = []
    values: list[float] = []


class Summary:
    def __init__(self, metrics, ...):
        ...
        self.metric_data = {
            metric: MetricData() for metric in metrics
        }

    ...
```

We could even use a `defaultdict` if we're feeling fancy.

Reply:
Indeed, this looks like a good idea.

@Daddelhai Please built upon this idea.